dango/types/src/oracle/wormhole.rs

```diff
@@ -0,0 +1,222 @@
+    pub emitter_address: [u8; 32],
+    pub payload: Vec<u8>,
+}
+
+impl WormholeVaa {
+    pub const HEADER_LEN: usize = 6;
+    pub const SIGNATURE_LEN: usize = 65;
+
+    pub fn new<T>(raw_bytes: T) -> anyhow::Result<Self>
+    where
+        T: Into<Vec<u8>>,
+    {
+        let mut bytes = BytesAnalyzer::new(raw_bytes.into());
+
+        let version = bytes.next_u8();
+        let guardian_set_index = bytes.next_u32()?;
+        let len_signers = bytes.next_u8();
+
+        let signatures = (0..len_signers)
+            .map(|_| {
+                let index = bytes.next_u8();
+                let signature = bytes.next_bytes::<{ WormholeVaa::SIGNATURE_LEN }>()?;
+                Ok((index, GuardianSignature::new(signature)?))
+            })
+            .collect::<anyhow::Result<BTreeMap<u8, GuardianSignature>>>()?;
+
+        // save some gas in wasm32
+        #[cfg(not(target_arch = "wasm32"))]
+        let hash = bytes.deref().keccak256().keccak256();
+        #[cfg(target_arch = "wasm32")]
+        let hash = Hash256::from_inner(
+            grug::ExternalApi.keccak256(&grug::ExternalApi.keccak256(bytes.deref())),
+        );
+
+        let timestamp = bytes.next_u32()?;
+        let nonce = bytes.next_u32()?;
+        let emitter_chain = bytes.next_u16()?;
+        let emitter_address = bytes.next_bytes::<32>()?;
+        let sequence = bytes.next_u64()?;
+        let consistency_level = bytes.next_u8();
+
+        Ok(WormholeVaa {
+            version,
+            guardian_set_index,
+            signatures,
+            hash,
+            timestamp,
+            nonce,
+            emitter_chain,
+            emitter_address,
+            sequence,
+            consistency_level,
+            payload: bytes.consume(),
+        })
+    }
+
+    pub fn verify(
+        self,
+        storage: &dyn Storage,
+        api: &dyn Api,
+        block: BlockInfo,
+        guardian_set: Map<u32, GuardianSetInfo>,
+    ) -> anyhow::Result<()> {
+        ensure!(self.version == 1, "Invalid VAA version");
+
+        let guardian_set = guardian_set.load(storage, self.guardian_set_index)?;
+
+        ensure!(
+            guardian_set.expiration_time == 0
+                || guardian_set.expiration_time as u128 > block.timestamp.into_inner().into_inner(),
+            "Guardian set expired"
+        );
+
+        ensure!(
+            guardian_set.quorum() <= self.signatures.len(),
+            "Not enough signatures"
+        );
+
+        for (index, sign) in self.signatures {
+            let pk =
+                api.secp256k1_pubkey_recover(&self.hash, &sign.signature, sign.id_recover, true)?;
+
+            let affine_point_option =
+                AffinePoint::from_encoded_point(&EncodedPoint::from_bytes(pk)?);
+            let affine_point = if affine_point_option.is_some().into() {
+                affine_point_option.unwrap()
+            } else {
+                bail!("Encoded point not on the curve");
+            };
```

Comment:
What you're doing here is first recover the pubkey as compressed (see the `false`) and then use the k256 library to decompress it. Instead you can just recover it as decompressed.